mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::shared_mempool::types::{PeerInfo, PeerSyncState};
+use libra_config::config::UpstreamConfig;
+use libra_types::PeerId;
+use rand::seq::IteratorRandom;
+use std::sync::Mutex;
+
+pub(crate) struct PeerManager {
+    upstream_config: UpstreamConfig,
+    peer_info: Mutex<PeerInfo>,
+    min_broadcast_recipient_count: usize,
+}
+
+impl PeerManager {
+    pub fn new(upstream_config: UpstreamConfig, min_broadcast_recipient_count: usize) -> Self {
+        let peer_info = Mutex::new(PeerInfo::new());
+        Self {
+            upstream_config,
+            peer_info,
+            min_broadcast_recipient_count,
+        }
+    }
+
+    pub fn add_peer(&self, network_id: PeerId, peer_id: PeerId) {
+        if self.is_upstream_peer(network_id, peer_id) {
+            println!("adding new peer");
+            self.peer_info
+                .lock()
+                .expect("failed to acquire peer info lock")
+                .entry(peer_id)
+                .or_insert(PeerSyncState {
+                    timeline_id: 0,
+                    is_alive: true,
+                    network_id,
+                })
+                .is_alive = true;
+        }
+    }
+
+    pub fn disable_peer(&self, peer_id: PeerId) {
+        if let Some(state) = self
+            .peer_info
+            .lock()
+            .expect("failed to acquire peer info lock")
+            .get_mut(&peer_id)
+        {
+            state.is_alive = false;
+        }
+    }
+
+    // pick_peers
+    // picks peers to broadcast to. Will first try to pick *all* preferred peers (= upstream as defined by config)
+    // if preferred upstream peers < `k = min_broadcast_recipient_count`, pick fallback peers to fill up k
+    pub fn pick_peers(&self) -> PeerInfo {
```

Comment:
I'm not sure if that's exactly a k-policy. So you're making sure that you sent at least to k fallback peers, but you still remove it locally once any peer acked. I thought idea is to drop it locally only once you received enough acks  
Or maybe it's a part of followup diff?

Reply:
yup, will handle ACK-side of k-policy in later PR